rir/src/compiler/native/lower_llvm.cpp

```diff
@@ -2454,16 +2508,133 @@ bool LowerFunctionLLVM::tryCompile() {
 
             case Tag::Extract2_1D: {
                 auto extract = Extract2_1D::Cast(i);
-                auto vector = loadSxp(i, extract->vec());
-                auto idx = loadSxp(i, extract->idx());
+                auto resultRep = representationOf(i);
+                // TODO: Extend a fastPath for generic vectors.
+                if (extract->vec()->type.isA(PirType::num().notObject()) &&
+                    extract->idx()->type.isScalar()) {
+                    auto fallback = BasicBlock::Create(C, "", fun);
+                    auto hit = BasicBlock::Create(C, "", fun);
+                    auto hit2 = BasicBlock::Create(C, "", fun);
+                    auto hit3 = BasicBlock::Create(C, "", fun);
+                    auto done = BasicBlock::Create(C, "", fun);
+                    
+                    llvm::Value* vector = load(i, extract->vec());
+                    llvm::Value* res =
+                        builder.CreateAlloca(representationOf(i));
+                    llvm::Value* index = nullptr;
+
+                    if (extract->vec()->type.maybeObj()) {
+                        auto rNil = constant(R_NilValue, t::SEXP);
+                        auto vectorhasAttr =
+                            builder.CreateICmpEQ(attr(vector), rNil);
+                        builder.CreateCondBr(vectorhasAttr, hit, fallback);
+                        builder.SetInsertPoint(hit);
+                    }
+
+                    if (representationOf(extract->idx()) !=
+                        Representation::Sexp) {
+                        if (representationOf(extract->idx()) ==
+                            Representation::Real) {
+                            index = builder.CreateTrunc(
+                                load(i, extract->idx(), Representation::Real),
+                                t::i64);
+                        } else {
+                            index = builder.CreateZExt(
+                                load(i, extract->idx(),
+                                     Representation::Integer),
+                                t::i64);
+                        }
+                    } else {
+                        // Maybe not needed if we know it is scalar?
```

Comment:
yes, isScalar() corresponds to the R level IS_SIMPLE_SCALAR macro. so actually isScalar => !maybeObject(). (If you see a type that is scalar but also maybeObject, that sounds wrong. We should probably have this as an assert in the types).